File: apps/desktop/src-tauri/src/model/manifest.rs
use serde_derive::Deserialize;
use serde_derive::Serialize;
use serde_json::Value;
use std::path::PathBuf;

pub const MANIFEST_FILE_NAME: &str = "package.json";

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum IconType {
    #[default]
    Iconify,
    AssetPath,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum TitleBarStyle {
    #[default]
    Visible,
    Transparent,
    Overlay,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum WindowTheme {
    #[default]
    Light,
    Dark,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum OSPlatform {
    Windows,
    MacOS,
    Linux,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Icon {
    pub icon: String,
    #[serde(rename = "type")]
    pub type_field: IconType,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Permissions {
    ClipboardRead,
    ClipboardWrite,
    FsHome,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct JarvisExtManifest {
    pub name: String,
    pub short_description: String,
    pub long_description: String,
    pub identifier: String,
    pub icon: Icon,
    pub demo_images: Vec<Value>,
    pub permissions: Option<Vec<Permissions>>,
    pub ui_cmds: Vec<UiCmd>,
    #[serde(default = "Vec::new")]
    pub inline_cmds: Vec<InlineCmd>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExtPackageJson {
    pub name: String,
    pub version: String,
    pub jarvis: JarvisExtManifest,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExtPackageJsonExtra {
    pub name: String,
    pub version: String,
    pub jarvis: JarvisExtManifest,
    // extra fields
    pub ext_path: PathBuf,
    pub ext_folder_name: String,
}

impl ExtPackageJsonExtra {
    pub fn from(manifest: ExtPackageJson, ext_path: PathBuf) -> Self {
        Self {
            name: manifest.name,
            version: manifest.version,
            jarvis: manifest.jarvis,
            ext_folder_name: ext_path.file_name().unwrap().to_str().unwrap().to_string(),
            ext_path,
        }
    }
}

impl ExtPackageJson {
    pub fn load(manifest_path: PathBuf) -> anyhow::Result<Self> {
        let manifest_str = std::fs::read_to_string(manifest_path).unwrap();
        Ok(serde_json::from_str(&manifest_str)?)
    }
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WindowConfig {
    pub center: Option<bool>,
    pub x: Option<u32>,
    pub y: Option<u32>,
    pub width: Option<u32>,
    pub height: Option<u32>,
    pub min_width: Option<u32>,
    pub min_height: Option<u32>,
    pub max_width: Option<u32>,
    pub max_height: Option<u32>,
    pub resizable: Option<bool>,
    pub title: Option<String>,
    pub fullscreen: Option<bool>,
    pub focus: Option<bool>,
    pub transparent: Option<bool>,
    pub maximized: Option<bool>,
    pub visible: Option<bool>,
    pub decorations: Option<bool>,
    pub always_on_top: Option<bool>,
    pub always_on_bottom: Option<bool>,
    pub content_protected: Option<bool>,
    pub skip_taskbar: Option<bool>,
    pub shadow: Option<bool>,
    pub theme: Option<WindowTheme>,
    pub title_bar_style: Option<TitleBarStyle>,
    pub hidden_title: Option<bool>,
    pub tabbing_identifier: Option<String>,
    pub maximizable: Option<bool>,
    pub minimizable: Option<bool>,
    pub closable: Option<bool>,
    pub parent: Option<String>,
    pub visible_on_all_workspaces: Option<bool>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UiCmd {
    pub main: String,
    pub name: String,
    pub description: Option<String>,
    pub dev_main: String,
    pub window: Option<WindowConfig>,
    pub cmds: Vec<Cmd>,
    pub platforms: Option<Vec<OSPlatform>>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Cmd {
    #[serde(rename = "type")]
    pub type_field: String,
    pub value: String,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InlineCmd {
    pub main: String,
    pub name: String,
    pub description: Option<String>,
    pub cmds: Vec<Cmd>,
    pub platforms: Option<Vec<OSPlatform>>,
}

// generate test
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_load_extension_manifest() {
        // this test relies on submodule
        let manifest_paths = vec![
            "../../../vendors/extensions/extensions/download-twitter-video/package.json",
            "../../../vendors/extensions/extensions/jwt/package.json",
            "../../../vendors/extensions/extensions/ip-info/package.json",
            "../../../vendors/extensions/extensions/qrcode/package.json",
        ];
        for manifest_path in manifest_paths {
            let manifest_str = std::fs::read_to_string(manifest_path).unwrap();
            let _: ExtPackageJson = serde_json::from_str(&manifest_str).unwrap();
        }
    }
}

File: packages/api/src/models/manifest.ts
import { FsPermissionSchema } from "tauri-api-adapter/permissions"
import {
	array,
	boolean,
	enum_,
	literal,
	nullable,
	number,
	object,
	optional,
	string,
	union,
	type InferOutput
} from "valibot"
import {
	AllKunkunPermission,
	FsPermissionScopedSchema,
	KunkunFsPermissionSchema,
	KunkunManifestPermission,
	OpenPermissionScopedSchema,
	ShellPermissionScopedSchema
} from "../permissions"
import { Icon } from "./icon"

export enum OSPlatformEnum {
	linux = "linux",
	macos = "macos",
	windows = "windows"
}

export const OSPlatform = enum_(OSPlatformEnum)
const allPlatforms = Object.values(OSPlatformEnum)
export const TriggerCmd = object({
	type: union([literal("text"), literal("regex")]),
	value: string()
})
export type TriggerCmd = InferOutput<typeof TriggerCmd>
export enum TitleBarStyleEnum {
	"visible" = "visible",
	"transparent" = "transparent",
	"overlay" = "overlay"
}
export const TitleBarStyle = enum_(TitleBarStyleEnum)
// JS new WebViewWindow only accepts lowercase, while manifest loaded from Rust is capitalized. I run toLowerCase() on the value before passing it to the WebViewWindow.
// This lowercase title bar style schema is used to validate and set the type so TypeScript won't complaint
// export const TitleBarStyleAllLower = z.enum(["visible", "transparent", "overlay"]);
// export type TitleBarStyleAllLower = z.infer<typeof TitleBarStyleAllLower>;
export const WindowConfig = object({
	center: optional(nullable(boolean())),
	x: optional(nullable(number())),
	y: optional(nullable(number())),
	width: optional(nullable(number())),
	height: optional(nullable(number())),
	minWidth: optional(nullable(number())),
	minHeight: optional(nullable(number())),
	maxWidth: optional(nullable(number())),
	maxHeight: optional(nullable(number())),
	resizable: optional(nullable(boolean())),
	title: optional(nullable(string())),
	fullscreen: optional(nullable(boolean())),
	focus: optional(nullable(boolean())),
	transparent: optional(nullable(boolean())),
	maximized: optional(nullable(boolean())),
	visible: optional(nullable(boolean())),
	decorations: optional(nullable(boolean())),
	alwaysOnTop: optional(nullable(boolean())),
	alwaysOnBottom: optional(nullable(boolean())),
	contentProtected: optional(nullable(boolean())),
	skipTaskbar: optional(nullable(boolean())),
	shadow: optional(nullable(boolean())),
	// theme: optional(nullable(union([literal("light"), literal("dark")]))), // changing theme of one window will change theme of all windows
	titleBarStyle: optional(nullable(TitleBarStyle)),
	hiddenTitle: optional(nullable(boolean())),
	tabbingIdentifier: optional(nullable(string())),
	maximizable: optional(nullable(boolean())),
	minimizable: optional(nullable(boolean())),
	closable: optional(nullable(boolean())),
	parent: optional(nullable(string())),
	visibleOnAllWorkspaces: optional(nullable(boolean()))
})
export type WindowConfig = InferOutput<typeof WindowConfig>
export const CustomUiCmd = object({
	main: string("HTML file to load, e.g. dist/index.html"),
	description: optional(nullable(string("Description of the Command"), ""), ""),
	devMain: string("URL to load in development to support live reload, e.g. http://localhost:5173/"),
	name: string("Name of the command"),
	window: optional(nullable(WindowConfig)),
	cmds: array(TriggerCmd, "Commands to trigger the UI"),
	platforms: optional(
		nullable(
			array(OSPlatform, "Platforms available on. Leave empty for all platforms."),
			allPlatforms
		),
		allPlatforms
	)
})
export type CustomUiCmd = InferOutput<typeof CustomUiCmd>

export const TemplateUiCmd = object({
	main: string(),
	name: string(),
	description: optional(nullable(string("Description of the Command"), ""), ""),
	cmds: array(TriggerCmd),
	platforms: optional(
		nullable(
			array(OSPlatform, "Platforms available on. Leave empty for all platforms."),
			allPlatforms
		),
		allPlatforms
	)
})
export type TemplateUiCmd = InferOutput<typeof TemplateUiCmd>

export const KunkunExtManifest = object({
	name: string("Name of the extension (Human Readable)"),
	shortDescription: string("Description of the extension (Will be displayed in store)"),
	longDescription: string("Long description of the extension (Will be displayed in store)"),
	identifier: string(
		"Unique identifier for the extension, must be the same as extension folder name"
	),
	icon: Icon,
	permissions: array(
		union([
			KunkunManifestPermission,
			FsPermissionScopedSchema,
			OpenPermissionScopedSchema,
			ShellPermissionScopedSchema
		]),
		"Permissions Declared by the extension. e.g. clipboard-all. Not declared APIs will be blocked."
	),
	demoImages: array(string("Demo images for the extension")),
	customUiCmds: array(CustomUiCmd, "Custom UI Commands"),
	templateUiCmds: array(TemplateUiCmd, "Template UI Commands")
})
export type KunkunExtManifest = InferOutput<typeof KunkunExtManifest>

export const ExtPackageJson = object({
	name: string("Package name for the extension (just a regular npm package name)"),
	version: string("Version of the extension"),
	kunkun: KunkunExtManifest,
	files: array(string("Files to include in the extension. e.g. ['dist']"))
})
export type ExtPackageJson = InferOutput<typeof ExtPackageJson>

/**
 * Extra fields for ExtPackageJson
 * e.g. path to the extension
 */
export const ExtPackageJsonExtra = object({
	...ExtPackageJson.entries,
	...{
		extPath: string(),
		extFolderName: string()
	}
})

export type ExtPackageJsonExtra = InferOutput<typeof ExtPackageJsonExtra>
